docs/content/en/docs/user-guide/configuration-reference.md

```diff
@@ -467,3 +468,26 @@ Note: By default, the sum of traffic is rounded to 100. If both `primary` and `c
 
 ### Percentage
 A wrapper of type `int` to represent percentage data. Basically, you can pass `10` or `"10"` or `10%` and they will be treated as `10%` in PipeCD.
+
+### KubernetesResourcePatch
+
+| Field | Type | Description | Required |
+|-|-|-|-|
+| target | [KubernetesResourcePatchTarget](/docs/user-guide/configuration-reference/#kubernetesresourcepatchtarget) | Which manifest, which field will be the target of patch operations. | Yes |
+| yamlOps | [][KubernetesResourcePatchYAMLOp](/docs/user-guide/configuration-reference/#kubernetesresourcepatchyamlop) | List of yaml operations should be applied to the above target. | Yes |
+
+### KubernetesResourcePatchTarget
+
+| Field | Type | Description | Required |
+|-|-|-|-|
+| kind | string | The resource kind. e.g. `ConfigMap` | Yes |
+| name | string | The resource name. e.g. `config-map-name` | Yes |
+| field | string | A string field whose content will be the target of patch operations. Empty means the whole manifest will be the target. e.g. `$.data.envoy-config` | No |
+
+### KubernetesResourcePatchYAMLOp
+
+| Field | Type | Description | Required |
+|-|-|-|-|
+| op | string | The operation type. This must be one of `replace`, `add`, `remove`. Default is `replace`. | No |
+| path | string | The path string pointing to the manipulated field. e.g. `$.foo.array[0].bar` | Yes |
+| value | string | The value string whose content will be used as new value for the field. | No |
```

Comment:
> I'm gonna say that people will wonder why both `yamlOps` and `path` are required. It can be confusing as it looks only for YAML. 

Sure. `yamlOps` should not be required. Let me fix this.

> Also in `KubernetesResourcePatchYAMLOp`, `op` and `value` are common settings no matter that is a `yaml` field (like envoy-config), so how about making it versatile

With the current way, we can have `yamlOps`, `jsonOps`, `textOps` or any `...Ops` we want and the good thing is that they can have their own definition. I would say `op`, `value`, `yamlField` can be good for `yamlOps` but for `textOps` by using regex they could not fit, we may need totally different ones.
Because of that, I am thinking the current way is more versatile and will not block future supports.

>  move field to `KubernetesResourcePatchYAMLOp`

Thb, I was thinking about this but decided to put it out. That struct/block is only for transforming/text operation.

I want to hear more thoughts about this configuration part from you. @nakabonne @khanhtc1202

Reply:
I get you guys' points, lets say I came with my own taste 😂 
```yaml
patches: 
  - target: 
       kind: ConfigMap
       name: foo
       field: $.data.envoy-config # omit for make whole file as target
       ops:
         - op: replace
            kind: yaml # can be omitted and default will be yaml
            path: $.foo.array[0].bar # required
            value: 50
```
And we could forget about `xxxOps` and type definitions such as `KubernetesResourcePatchYAMLOp`. Of course, in case the `textOps` for example have completely different fields, @nghialv 's suggestion is better (type for each) but I feel we don't have much difference between xxxOps type. 🤔

Reply:
That looks nice. 👍

Reply:
@khanhtc1202 looks a bit more complex to parse but nicer. One little thing that I want to wonder is the default should be `text`. And, here is what I'm kind of imaging about text ops.

```yaml
patches:
- target:
    kind: Deployment
    name: foo
    field: $.spec.template.spec.containers[0].env[0].value
  ops:
    - op: replace
      kind: text # default is text
      value: bar

- target:
    kind: ConfigMap
    name: long-text
    field: $.data.envoy-config
  ops:
    - op: replace
      kind: text
      regex: (.+)-[0-9a-zA-Z]+(-[0-9a-zA-Z]+)$ # optional
      value: bar

- target:
    kind: ConfigMap
    name: envoy-config
    field: $.data.envoy-config
  ops:
    - op: replace
      kind: yaml
      path: $.resources[0].virtual_hosts[0].routes[0].route.weighted_clusters.clusters[0].weight
      value: 50
```

Reply:
Thank you guys for your comments. I think finally we came up with the following versions:

1. `target.field` is optional (path, value are optional as well)

``` yaml
patches:
- target:
    kind: Deployment
    name: foo
  ops:
  - op: replace
    kind: yaml # default is yaml
    path: $.spec.template.spec.containers[0].env[0].value
    value: bar

- target:
    kind: ConfigMap
    name: envoy-config
    field: $.data.envoy-config
  ops:
  - op: replace
    kind: yaml
    path: $.resources[0].virtual_hosts[0].routes[0].route.weighted_clusters.clusters[0].weight
    value: 50
  - op: replace
    kind: yaml
    path: $.resources[0].virtual_hosts[0].routes[0].route.weighted_clusters.clusters[1].weight
    value: 50
```

2. `target.field` is considered as a mandatory field
 
``` yaml
patches:
- target:
    kind: Deployment
    name: foo
    field: $.spec.template.spec.containers[0].env[0].value
  ops:
    - op: replace
      kind: text # default is text
      value: bar

- target:
    kind: ConfigMap
    name: long-text
    field: $.data.long-text
  ops:
    - op: replace
      kind: text
      regex: (.+)-[0-9a-zA-Z]+(-[0-9a-zA-Z]+)$ # optional
      value: bar

- target:
    kind: ConfigMap
    name: envoy-config
    field: $.data.envoy-config
  ops:
    - op: replace
      kind: yaml
      path: $.resources[0].virtual_hosts[0].routes[0].route.weighted_clusters.clusters[0].weight
      value: 50
```

For the last one, it could be hard to use when we want to apply multiple operations for the same Kubernetes manifests.
For example, updating multiple environment variables for a Deployment manifest.

Reply:
For the first one, we can also merge `kind` and `op` into a new one

``` yaml
patches:
- target:
    kind: Deployment
    name: foo
  ops:
  - op: yaml-replace
    path: $.spec.template.spec.containers[0].env[0].value
    value: bar

- target:
    kind: ConfigMap
    name: envoy-config
    field: $.data.envoy-config
  ops:
  - op: yaml-replace
    path: $.resources[0].virtual_hosts[0].routes[0].route.weighted_clusters.clusters[0].weight
    value: 50
  - op: yaml-replace
    path: $.resources[0].virtual_hosts[0].routes[0].route.weighted_clusters.clusters[1].weight
    value: 50
```

`op` can be one of `yaml-replace`, `yaml-add`, `json-replace`, `text-regex`...

Reply:
Hmm, well, got harder to understand for newbies.

Reply:
I agree with
- making it possible to apply multiple operations for the same Kubernetes manifests.
- merging kind and op (like `yaml-replace`)

Reply:
But I think all target field specify a single value, not a YAML document. Here, we have to completely distinguish a YAML document and a single string in the YAML format. I mean

This is a YAML document

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: foo
  labels:
```

On the other hand, this is a single string in the YAML format. I'd say we shouldn't treat these in the same way.

```yaml
data:
  envoy-config: |-
    admin:
      address:
        socket_address:
          address: 0.0.0.0
          port_value: 9095
```

Reply:
What I'm trying to say is that we shouldn't handle these as the same YAML kind.

```yaml
  - op: replace
    kind: yaml # default is yaml
    path: $.spec.template.spec.containers[0].env[0].value
    value: bar
```

```yaml
  - op: replace
    kind: yaml
    path: $.resources[0].virtual_hosts[0].routes[0].route.weighted_clusters.clusters[0].weight
    value: 50
```

Reply:
Considering them, how about something like this?

```yaml
patches:
- resource:
    kind: Deployment
    name: foo
    targets:
      - field: $.spec.template.spec.containers[0].env[0].value
        op: replace # optional: default is replace
        value: bar
      - field: $.spec.template.spec.containers[0].env[0].name
        op: replace
        value: env1
- resource:
    kind: ConfigMap
    name: envoy-config
    targets:
      - field: $.data.envoy-config
        op: yaml-replace
        path: $.resources[0].virtual_hosts[0].routes[0].route.weighted_clusters.clusters[0].weight
        value: 50
      - field: $.data.envoy-config
        op: yaml-replace
        path: $.resources[0].virtual_hosts[0].routes[0].route.weighted_clusters.clusters[1].weight
        value: 50
```

Reply:
I feel merging the "operation" thing into "target" as above is not a good idea.

And it is a little strange to see both "field" and "path" in the same struct.

> all fields are required

maybe you meant "optional"

``` yaml
patches:
- target:
    kind: Deployment
    name: foo
  ops:
  - op: yaml-replace
    path: $.spec.template.spec.containers[0].env[0].value
    value: bar
  - op: yaml-add
    path: $.spec.template.spec.containers[0].env
    value: "new: 1"

- target:
    kind: ConfigMap
    name: envoy-config
    field: $.data.envoy-config
  ops:
  - op: yaml-replace
    path: $.resources[0].virtual_hosts[0].routes[0].route.weighted_clusters.clusters[0].weight
    value: 50
  - op: yaml-replace
    path: $.resources[0].virtual_hosts[0].routes[0].route.weighted_clusters.clusters[1].weight
    value: 50
```

At least in this way, the target and the text operation are separate.
The target will be considered as the way to decide what will be the input of the next operations.
The input can be a string that contains the whole "YAML" data or a sub-string of the manifest.

Reply:
I know what you want to do. It would be quite helpful if we could add entities to a YAML sequence like `yaml-add`.

But anyway I feel quite it's difficult to understand that we handle a YAML document and a single string in the YAML format in the same way.

Reply:
I meant all fields other than `op` are required. But not significant at all, never mind.

Reply:
Why not consider a single string in the YAML format as an edge case? Of course, it's not an edge case because modifying ConfigMap often happens.

Reply:
How about something like this?

```yaml
patches:
- kind: Deployment
  name: foo
  fields:
    - field: $.spec.template.spec.containers[0].env[0].value
      ops:
        - op: replace
          value: bar
    - field: $.spec.template.spec.containers[0].env
      ops:
        - op: add
          value: "new: 1"

- kind: ConfigMap
  name: envoy-config
  fields:
    - field: $.data.envoy-config
      ops:
        - op: yaml-replace
          path: $.resources[0].virtual_hosts[0].routes[0].route.weighted_clusters.clusters[0].weight
          value: 50
        - op: yaml-replace
          path: $.resources[0].virtual_hosts[0].routes[0].route.weighted_clusters.clusters[1].weight
          value: 50
```

Reply:
I think in both ways, we are taking `single string`/`handling a specific field` as an edge case.
Just the way to represent it.

In this way https://github.com/pipe-cd/pipe/pull/2333#discussion_r682207323, that edge case can be enabled by specifying `target.field` field.

In this way https://github.com/pipe-cd/pipe/pull/2333#discussion_r681626182, that edge case can be enabled by introducing `path` and more `op`s.

Reply:
Multiple fields in a manifest can be modified, and multiple operations can be done for a field. Besides, for me it's more easy to understand because the structure is nested from (Manifest -> Field -> Ops (it can be different depending on value type). Like for instance, if the value specified by `field` is just a YAML document (not a string in YAML format like envoy-config), you don't have to specify `path`.

This can:
- apply multiple operations for the same Kubernetes manifests

Reply:
>I think in both ways, we are taking single string/handling a specific field as an edge case.

You're right.

Reply:
In this way https://github.com/pipe-cd/pipe/pull/2333#discussion_r682228025, we will be heavily dependent on the `field`.
All operations are just for the specified fields. And we may lose one of our powerful weapons: `text-regex` on the whole manifest.

And another problem can be raised when using `yaml-replace` for `Deployment` instead of `replace`.

Reply:
>And we may lose one of our powerful weapons: text-regex

This is exactly what I've been thinking about. That's exactly an extreme edge case and when does it needed? Could you please show me some examples?

>And another problem can be raised when using yaml-replace for Deployment instead of replace.

Sorry, could you also tell me some examples?

Reply:
> That's exactly an extreme edge case and when does it needed? Could you please show me some examples?

Like `event-watcher` we may face that need in the future. For example, the user wants to suffix all matching sub-string in the manifest with `canary` string.

> Sorry, could you also tell me some examples?

I meant it could be confused about where to use `replace` and where to use `yaml-replace`. Should we allow using `yaml-replace` to patch `Deployment` in your example or not.

Reply:
Thank you for your examples!

>For example, the user wants to suffix all matching sub-string in the manifest with canary string.

I see. I think the target file of Event Watcher can be sometimes not formatted, but for k8s manifests, it is definitely formatted. Therefore, it seems to be no problem that the config structure depends on the `field`.

Making it flexible comes with complexity. In the case you showed me, you just have to add all fields to `fields`. It's tedious but this is an edge case. I think it doesn't need to make the configuration more complicated for just a helper feature.

Reply:
>I meant it could be confused about where to use replace and where to use yaml-replace.

Again, it's a special case that a value string is formatted in YAML. So basically we just use `replace`, `add`, and `remove`. I think just explaining to users like this is enough: "Only if the value of the specified `field` is a string in YAML format, a special op called `yaml-replace` is available."

Reply:
> I think it doesn't need to make the configuration more complicated for just a helper feature.

Yeah. I totally agree about this. If we had a better solution that is easier to understand than https://github.com/pipe-cd/pipe/pull/2333#discussion_r681605175,  we should apply it. But IMO I still don't see that https://github.com/pipe-cd/pipe/pull/2333#discussion_r682228025 and https://github.com/pipe-cd/pipe/pull/2333#discussion_r681626182 are easy to understand.

Reply:
> Again, it's a special case that a value string is formatted in YAML. So basically we just use replace, add, and remove. I think just explaining to users like this is enough: "Only if the value of the specified field is a string in YAML format, a special op called yaml-replace is available."

Yeah. So why not https://github.com/pipe-cd/pipe/pull/2333#discussion_r682207323. `field` is specified for only edge case and no additional explanation is needed.

Reply:
Sorry, the only thing I want to argue you is to stop treating a YAML document and a string formatted in YAML in the same way.

Reply:
Do you that that is a bad thing? And that is bad for us or bad for our users?

Reply:
For our users

Reply:
Honestly I can still be confused that two types of YAML are mixed in the same context.

Reply:
The purposes of all examples that I showed are just to separate them.

Reply:
I believe first we have to specify `field`, then based on that, specify ops, like this:

```yaml
patches:
- target:
    kind: Deployment
    name: foo
    field: $.spec.template.spec.containers[0].env[0].value
  ops:
  - op: replace
    value: bar

- target:
    kind: Deployment
    name: foo
    field: $.spec.template.spec.containers[0].env
  ops:
  - op: add
    value: "new: 1"

- target:
    kind: ConfigMap
    name: envoy-config
    field: $.data.envoy-config
  ops:
  - op: yaml-replace
    path: $.resources[0].virtual_hosts[0].routes[0].route.weighted_clusters.clusters[0].weight
    value: 50
  - op: yaml-replace
    path: $.resources[0].virtual_hosts[0].routes[0].route.weighted_clusters.clusters[1].weight
    value: 50
```

But as you can see, we have to repeat the same manifest to operate for multiple fields in the same manifest. That's why I suggested such a thing.

Reply:
If I understand right, it's way too flexible that yaml-replace can be used for a YAML document and a string formatted in YAML.

Reply:
@khanhtc1202 Do you have any better idea that came with your own taste again?

Reply:
@khanhtc1202 To summarize it, @nghialv is suggesting something like [this](https://github.com/pipe-cd/pipe/pull/2333#discussion_r682207323) and I'm suggesting a kind of like [this](https://github.com/pipe-cd/pipe/pull/2333#discussion_r682228025).

Reply:
Sorry, it took me time to catch up on all the above comments of this discussion 👀  If I get right, @nakabonne worries about the confusion in the usage of field and path are not good for our users and it's inconvenient if we want to apply multiple ops in the same manifest with different fields, right? I think it is the place where we should decide what will be used frequently to make choices of what should have higher priority 🤔

Reply:
> it's inconvenient if we want to apply multiple ops in the same manifest with different fields, right?

No, I don't think this is his worry.

Reply:
Let's me put both of the final candicates here for comparison.

Reply:
as I can see, we have a target-based option and a field-based option. If we keep going as we do with the secret management feature, the target-based option has a bit advantage over the field-based. Another thing that concerned me is what if the ops don't require `field` value (regex case), what will the configuration be in that case? 🤔

Reply:
>No, I don't think this is his worry.

Yeah, what I'm worried about is just treating a YAML document and a string formatted in YAML in the same way (sorry if it's difficult to understand, I don't know how to say that exactly 😓 ).

Reply:
1. Let's call it as `number 1`

- using `target` to show which text should be used in the input of all `ops` and `target.field` to specify a sub-text
- `op` can be one of `yaml-replace`, `yaml-add`, `text-regex`...

``` yaml
patches:
- target:
    kind: Deployment
    name: foo
  ops:
  - op: yaml-replace
    path: $.spec.template.spec.containers[0].env[0].value
    value: bar
  - op: yaml-add
    path: $.spec.template.spec.containers[0].env
    value: "new: 1"

- target:
    kind: ConfigMap
    name: envoy-config
    field: $.data.envoy-config
  ops:
  - op: yaml-replace
    path: $.resources[0].virtual_hosts[0].routes[0].route.weighted_clusters.clusters[0].weight
    value: 50
  - op: yaml-replace
    path: $.resources[0].virtual_hosts[0].routes[0].route.weighted_clusters.clusters[1].weight
    value: 50
```

2. Let's call it as `number 2`

- Introducing `field`
- ops can be one of `replace`, `add`, `yaml-replace`, `yaml-add`...

```yaml
patches:
- kind: Deployment
  name: foo
  fields:
    - field: $.spec.template.spec.containers[0].env[0].value
      ops:
        - op: replace
          value: bar
    - field: $.spec.template.spec.containers[0].env
      ops:
        - op: add
          value: "new: 1"

- kind: ConfigMap
  name: envoy-config
  fields:
    - field: $.data.envoy-config
      ops:
        - op: yaml-replace
          path: $.resources[0].virtual_hosts[0].routes[0].route.weighted_clusters.clusters[0].weight
          value: 50
        - op: yaml-replace
          path: $.resources[0].virtual_hosts[0].routes[0].route.weighted_clusters.clusters[1].weight
          value: 50
```

Reply:
This is nicer.

```yaml
patches:
- target:
    kind: Deployment
    name: foo
    documentRoot: $ # default is $
  ops:
  - op: yaml-replace
    path: $.spec.template.spec.containers[0].env[0].value
    value: bar
  - op: yaml-add
    path: $.spec.template.spec.containers[0].env
    value: "new: 1"

- target:
    kind: ConfigMap
    name: envoy-config
    documentRoot: $.data.envoy-config
  ops:
  - op: yaml-replace
    path: $.resources[0].virtual_hosts[0].routes[0].route.weighted_clusters.clusters[0].weight
    value: 50
  - op: yaml-replace
    path: $.resources[0].virtual_hosts[0].routes[0].route.weighted_clusters.clusters[1].weight
    value: 50
```

Reply:
I'm okay with this 🙌

Reply:
Thank you guys. 💯